File: android/app/src/main/java/com/kurt/bgmate/data/remote/llm/GeminiLlmClient.kt
package com.kurt.bgmate.data.remote.llm

import android.util.Log
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.flowOn
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okio.IOException
import org.json.JSONArray
import org.json.JSONObject

class GeminiLlmClient(
    private val okHttpClient: OkHttpClient,
    private val apiKey: String
) : LlmClient {

    companion object {
        private const val TAG = "GeminiLlmClient"
        private const val BASE_URL =
            "https://generativelanguage.googleapis.com/v1beta/models/gemini-2.5-flash"
        private const val GENERATE_CONTENT_URL = "$BASE_URL:generateContent"
        private const val STREAM_GENERATE_CONTENT_URL = "$BASE_URL:streamGenerateContent?alt=sse"

        // Gemini uses "model" instead of "assistant" for the assistant role
        private fun mapRole(role: String): String = when (role) {
            "assistant" -> "model"
            else -> role
        }
    }

    override suspend fun complete(request: LlmRequest): LlmResponse = withContext(Dispatchers.IO) {
        val body = buildRequestBody(request)

        val httpRequest = Request.Builder()
            .url(GENERATE_CONTENT_URL)
            .addHeader("x-goog-api-key", apiKey)
            .addHeader("content-type", "application/json")
            .post(body.toString().toRequestBody("application/json".toMediaType()))
            .build()

        val response = okHttpClient.newCall(httpRequest).execute()
        val responseBody = response.body.string()

        if (!response.isSuccessful) {
            throw IOException("Gemini API 오류 ${response.code}: $responseBody")
        }
        val text = extractTextFromResponse(responseBody)
        LlmResponse(text)
    }

    override fun stream(request: LlmRequest): Flow<String> = flow {
        val body = buildRequestBody(request)

        val httpRequest = Request.Builder()
            .url(STREAM_GENERATE_CONTENT_URL)
            .addHeader("x-goog-api-key", apiKey)
            .addHeader("content-type", "application/json")
            .post(body.toString().toRequestBody("application/json".toMediaType()))
            .build()

        val response = okHttpClient.newCall(httpRequest).execute()

        if (!response.isSuccessful) {
            throw IOException("Gemini API 오류 ${response.code}")
        }

        response.body.source().use { source ->
            while (!source.exhausted()) {
                val line = source.readUtf8Line() ?: break

                if (!line.startsWith("data: ")) continue

                val data = line.removePrefix("data: ")

                val chunk = runCatching {
                    extractTextFromResponse(data)
                }.getOrDefault("")

                if (chunk.isNotEmpty()) emit(chunk)
            }
        }
    }.flowOn(Dispatchers.IO)

    private fun buildRequestBody(request: LlmRequest): JSONObject = JSONObject().apply {
        request.systemPrompt?.let { prompt ->
            put("systemInstruction", JSONObject().apply {
                put("parts", JSONArray().apply {
                    put(JSONObject().apply { put("text", prompt) })
                })
            })
        }

        put("contents", JSONArray().apply {
            request.messages.forEach { msg ->
                put(JSONObject().apply {
                    put("role", mapRole(msg.role))
                    put("parts", JSONArray().apply {
                        put(JSONObject().apply { put("text", msg.content) })
                    })
                })
            }
        })

        put("generationConfig", JSONObject().apply {
            put("maxOutputTokens", request.maxTokens)
        })
    }

    private fun extractTextFromResponse(responseBody: String): String {
        val root = JSONObject(responseBody)
        val candidates = root.optJSONArray("candidates") ?: return ""
        if (candidates.length() == 0) return ""

        val candidate = candidates.getJSONObject(0)
        val finishReason = candidate.optString("finishReason")
        if (finishReason == "MAX_TOKENS") {
            Log.w(TAG, "Gemini 응답이 MAX_TOKENS로 중단되었습니다. maxTokens 값을 높이세요.")
        }

        val content = candidate.optJSONObject("content") ?: return ""
        val parts = content.optJSONArray("parts") ?: return ""

        // gemini-2.5-flash는 사고(thinking) 파트를 포함할 수 있음.
        // thought=true 파트는 내부 추론이므로 제외하고 실제 응답만 추출.
        val sb = StringBuilder()
        for (i in 0 until parts.length()) {
            val part = parts.getJSONObject(i)
            if (!part.optBoolean("thought", false)) {
                sb.append(part.optString("text", ""))
            }
        }
        return sb.toString()
    }
}

File: flutter/lib/data/remote/ai/gemini_llm_client.dart
import 'dart:async';
import 'dart:convert';
import 'dart:developer' as dev;

import 'package:bgmate_flutter/data/remote/ai/llm_client.dart';
import 'package:bgmate_flutter/data/remote/ai/llm_request.dart';
import 'package:bgmate_flutter/data/remote/ai/llm_response.dart';
import 'package:dio/dio.dart';

class GeminiLlmClient implements LlmClient {
  final Dio _dio;
  final String _apiKey;

  static const _tag = 'GeminiLlmClient';
  static const _baseUrl =
      'https://generativelanguage.googleapis.com/v1beta/models/gemini-2.5-flash';
  static const _generateContentUrl = '$_baseUrl:generateContent';
  static const _streamGenerateContentUrl =
      '$_baseUrl:streamGenerateContent?alt=sse';

  GeminiLlmClient({required Dio dio, required String apiKey})
      : _dio = dio,
        _apiKey = apiKey;

  // Gemini는 assistant role을 "model"로 사용
  static String _mapRole(String role) =>
      role == 'assistant' ? 'model' : role;

  @override
  Future<LlmResponse> complete(LlmRequest request) async {
    final body = _buildRequestBody(request);
    dev.log('→ complete request: ${jsonEncode(body)}', name: _tag);

    try {
      final response = await _dio.post<Map<String, dynamic>>(
        _generateContentUrl,
        data: body,
        options: Options(
          headers: {
            'x-goog-api-key': _apiKey,
            'content-type': 'application/json',
          },
        ),
      );
      dev.log('← complete ${response.statusCode}: ${jsonEncode(response.data)}',
          name: _tag);
      return LlmResponse(text: _extractText(response.data ?? {}));
    } on DioException catch (e) {
      dev.log(
        '✗ complete error ${e.response?.statusCode}: ${e.response?.data}',
        name: _tag,
        error: e,
      );
      rethrow;
    }
  }

  @override
  Stream<String> stream(LlmRequest request) async* {
    final body = _buildRequestBody(request);
    dev.log('→ stream request: ${jsonEncode(body)}', name: _tag);

    final Response<ResponseBody> response;
    try {
      response = await _retryOnOverload(() => _dio.post<ResponseBody>(
        _streamGenerateContentUrl,
        data: body,
        options: Options(
          headers: {
            'x-goog-api-key': _apiKey,
            'content-type': 'application/json',
          },
          responseType: ResponseType.stream,
        ),
      ));
    } on DioException catch (e) {
      dev.log(
        '✗ stream error ${e.response?.statusCode}: ${e.response?.data}',
        name: _tag,
        error: e,
      );
      rethrow;
    }

    dev.log('← stream connected (status ${response.statusCode})', name: _tag);

    final lines = response.data!.stream
        .cast<List<int>>()
        .transform(utf8.decoder)
        .transform(const LineSplitter());

    await for (final line in lines) {
      dev.log('  SSE: $line', name: _tag);
      if (!line.startsWith('data: ')) continue;

      final data = line.substring('data: '.length);
      final chunk = _extractTextSafe(data);
      if (chunk.isNotEmpty) yield chunk;
    }
  }

  /// 503/429 응답 시 지수 백오프로 최대 [maxAttempts]회 재시도합니다.
  Future<T> _retryOnOverload<T>(
    Future<T> Function() call, {
    int maxAttempts = 3,
  }) async {
    for (var attempt = 1; attempt <= maxAttempts; attempt++) {
      try {
        return await call();
      } on DioException catch (e) {
        final status = e.response?.statusCode;
        final retryable = status == 503 || status == 429;
        if (!retryable || attempt == maxAttempts) rethrow;

        final delay = Duration(seconds: 1 << (attempt - 1)); // 1s, 2s, 4s
        dev.log(
          '↻ attempt $attempt failed ($status), retrying in ${delay.inSeconds}s…',
          name: _tag,
        );
        await Future<void>.delayed(delay);
      }
    }
    throw StateError('unreachable');
  }

  Map<String, dynamic> _buildRequestBody(LlmRequest request) {
    return {
      if (request.systemPrompt != null)
        'systemInstruction': {
          'parts': [
            {'text': request.systemPrompt},
          ],
        },
      'contents': [
        for (final msg in request.messages)
          {
            'role': _mapRole(msg.role),
            'parts': [
              {'text': msg.content},
            ],
          },
      ],
      'generationConfig': {'maxOutputTokens': request.maxTokens},
    };
  }

  String _extractText(Map<String, dynamic> root) {
    final candidates = root['candidates'] as List<dynamic>?;
    if (candidates == null || candidates.isEmpty) return '';

    final candidate = candidates[0] as Map<String, dynamic>;
    final content = candidate['content'] as Map<String, dynamic>?;
    final parts = content?['parts'] as List<dynamic>?;
    if (parts == null) return '';

    // gemini-2.5-flash는 사고(thinking) 파트를 포함할 수 있음.
    // thought=true 파트는 내부 추론이므로 제외하고 실제 응답만 추출.
    return parts
        .whereType<Map<String, dynamic>>()
        .where((part) => part['thought'] != true)
        .map((part) => part['text'] as String? ?? '')
        .join();
  }

  String _extractTextSafe(String responseBody) {
    try {
      final root = jsonDecode(responseBody) as Map<String, dynamic>;
      return _extractText(root);
    } catch (_) {
      return '';
    }
  }
}
